checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -364,24 +364,26 @@ private static int getGroupCount(Pattern p) {
     /**
      * Return the non-null groups in the argument.
      *
-     * @param p pattern to be analysed
-     * @return the groups that are non-null in the argument
+     * @param regexp pattern to be analysed.
+     * @param n number of capturing groups in the pattern.
+     * @return the groups that are non-null in the argument.
      */
-    private static List<Integer> getNonNullGroups(Pattern p) {
-        String regexp = p.pattern();
-        int n = getGroupCount(p);
+    public static List<Integer> getNonNullGroups(String regexp, int n) {
         List<Integer> nonNullGroups = new ArrayList<>();
-        for (int i = 0; i <= n; i++) {
+        for (int i = 1; i <= n; i++) {
             nonNullGroups.add(i);
         }
         ArrayDeque<Integer> openingIndices = new ArrayDeque<>();
         int group = 0;
         boolean squareBracketOpen = false;
         boolean escaped = false;
         boolean notCapturing = false;
+        boolean quoted = false;
         for (int i = 0; i < regexp.length(); i++) {
-            if (!escaped && !squareBracketOpen && regexp.charAt(i) == '(') {
-                if (i != regexp.length() - 1 && regexp.charAt(i + 1) == '?') {
+            if (!quoted && !escaped && !squareBracketOpen && regexp.charAt(i) == '(') {
```

Comment:
Why don't you need to add a `!quoted` test elsewhere, such as at the test for `')'`?  It would be helpful to document the relationship among the 4 boolean variables: what values can they have at the same time?

Reply:
I was just plugging in conditions as I discovered them, it became really messy, hard to understand, and easy to mess up. Now after reading the documentation for Pattern, I think I have covered everything and explained it nicely.